README.md
Docker webapp image
-------------------

[![Build Status](https://travis-ci.org/roquie/docker-php-webapp.svg?branch=master)](https://travis-ci.org/roquie/docker-php-webapp)
![Docker Automated build](https://img.shields.io/docker/automated/roquie/docker-php-webapp.svg)
![MicroBadger Size (tag)](https://img.shields.io/microbadger/image-size/roquie/docker-php-webapp.svg)
[![GitHub license](https://img.shields.io/github/license/roquie/docker-php-webapp.svg)](https://github.com/roquie/docker-php-webapp)

## Introduction

A classic container who runs NGINX and PHP-FPM together. For best performance I recommend
use my [swoole docker image](https://github.com/roquie/docker-swoole-webapp) or [RoadRunner](https://github.com/spiral/roadrunner)
if `swoole` extension not suitable for you.

Docker image includes PHP 7.3/7.4, NGINX, PHP-FPM and supervisord.

Kill features:
* Security-frendly because work as non root user.
* Fast start at run.
* Used the env variables for configure many parameters at start.
* Fully compatible with PaaS-solutions like as Heroku, Flynn, Deis Workflow, Dokku and many others.
* Internal port of nginx is taken from `$PORT` env variable.
* Used Supervisord written on Golang.
* Small image size (for Debian-based of course).

**Every week travis auto build the docker image and push to the registry.**

## ENV

```bash
ENV PORT=8080
ENV NGINX_WEBROOT="/srv/www"
ENV NGINX_WORKER_PROCESSES=auto # If not set, when image start, script automatically check how many of CPUs count contains within machine.
ENV NGINX_MAX_BODY_SIZE="2G"
ENV NGINX_BODY_TIMEOUT="120s"
ENV NGINX_HEADER_TIMEOUT="120s"
ENV NGINX_ACCESS_LOG="off"

ENV PHP_FPM_CLEAR_ENV="no"
ENV PHP_FPM_LOG_LEVEL="notice"

ENV PHP_ERROR_REPORTING="E_ALL & ~E_DEPRECATED & ~E_STRICT"
ENV PHP_DISPLAY_ERRORS="Off"
ENV PHP_SHORT_OPEN_TAG="Off"
ENV PHP_MEMORY_LIMIT="512M"
ENV PHP_POST_MAX_SIZE="128M"
ENV PHP_UPLOAD_MAX_FILESIZE="128M"
ENV PHP_MAX_EXECUTION_TIME="160"
ENV PHP_MAX_INPUT_TIME="120"
ENV PHP_DATE_TIMEZONE="UTC"
ENV PHP_EXPOSE="Off"
ENV PHP_OPCACHE_ENABLED="1"
ENV PHP_OPCACHE_SAVE_COMMENTS="0"
```

## Ports

* 8080

## Volumes

* /srv/www

## Usage example

```bash
FROM roquie/docker-php-webapp:7.4-1.5.0

COPY --chown=nginx:nginx . /srv/www
```

## Versions

* latest
* 7.4-latest or 7.4-{semver} 
* 7.3-latest

## License

Copyright (c) 2020 Roquie <roquie0@gmail.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.



CHANGELOG.md
## [1.5.0] - 02-06-2020
### Use unix sockets instead of localhost
- Unix sockets
- Reduced image size (I forgot delete `php7.4-dev` package) 
- `keepalive 22` -> `keepalive 30`
- Increased fastcgi buffer size.

## [1.4.0] - 02-06-2020
### PHP 7.4 support
- Opcache preload support thought `$PHP_OPCACHE_PRELOAD`
- Using semver in image tag name. 
- `opcache.file_update_protection=0` as default option instead of `2`.
- `7.4-1.4.0` tag

## [1.3.0] - 02-06-2020
### PHP 7.4 support
- Drop 7.2
- Fixes some bugs
- `7.4-beta1` tag

## [1.2.0] - 30-09-2019
### Bugfixes
- Fixed opcache env variable
- Fixed logs output (Use `PHP_FPM_LOG_LEVEL` for configure it)
- Add `bcmath` and `xmlreader` extensions.
- Commented `user` and `group` options in php-fpm config

## [1.1.0] - 10-02-2019
### PHP 7.3
- Fixed failed travis build (gpg tty problem)
- Support multiple PHP versions
- Add new docker tags `php7.2-latest` and `php7.3-latest`

## [1.0.1] - 02-10-2018
### Bugfix
- Fixed container quitting when user send `Ctrl+C` input

## [1.0.0] - 21-09-2018
### First stable version
- Drop varrick template engine because his doesn't work in kubernetes cluster
- Use more fastest and stable roquie/smalte template engine
- Reduce image size
- Added NGINX_ACCESS_LOG env variable

## [0.1.0] - 15-06-2018
### First version
- Created Docker image with thrully features.


conf/start.sh
#!/bin/bash

# CPU cores count.
count=$(grep processor /proc/cpuinfo | wc -l)

export PORT=${PORT:=8080}
export NGINX_WEBROOT=${NGINX_WEBROOT:="/srv/www"}
export NGINX_WORKER_PROCESSES=${NGINX_WORKER_PROCESSES:=${count}}
export NGINX_MAX_BODY_SIZE=${NGINX_MAX_BODY_SIZE:="2G"}
export NGINX_BODY_TIMEOUT=${NGINX_BODY_TIMEOUT:="120s"}
export NGINX_HEADER_TIMEOUT=${NGINX_HEADER_TIMEOUT:="120s"}
export NGINX_ACCESS_LOG=${NGINX_ACCESS_LOG:="off"}

export PHP_FPM_CLEAR_ENV=${PHP_FPM_CLEAR_ENV:="no"}
export PHP_FPM_LOG_LEVEL=${PHP_FPM_LOG_LEVEL:="notice"}

export PHP_ERROR_REPORTING=${PHP_ERROR_REPORTING:="E_ALL & ~E_DEPRECATED & ~E_STRICT"}
export PHP_DISPLAY_ERRORS=${PHP_DISPLAY_ERRORS:="Off"}
export PHP_SHORT_OPEN_TAG=${PHP_SHORT_OPEN_TAG:="Off"}
export PHP_MEMORY_LIMIT=${PHP_MEMORY_LIMIT:="512M"}
export PHP_POST_MAX_SIZE=${PHP_POST_MAX_SIZE:="128M"}
export PHP_UPLOAD_MAX_FILESIZE=${PHP_UPLOAD_MAX_FILESIZE:="128M"}
export PHP_MAX_EXECUTION_TIME=${PHP_MAX_EXECUTION_TIME:="160"}
export PHP_MAX_INPUT_TIME=${PHP_MAX_INPUT_TIME:="120"}
export PHP_DATE_TIMEZONE=${PHP_DATE_TIMEZONE:="UTC"}
export PHP_EXPOSE=${PHP_EXPOSE:="Off"}

# backward compatibility, $PHP_OPCACHE is deprecated.
if [[ "$PHP_OPCACHE" != "" ]]
then
    export PHP_OPCACHE_ENABLED="$PHP_OPCACHE"
    unset PHP_OPCACHE
fi

export PHP_OPCACHE_ENABLED=${PHP_OPCACHE_ENABLED:="1"}
export PHP_OPCACHE_PRELOAD=${PHP_OPCACHE_PRELOAD:=""}
export PHP_OPCACHE_SAVE_COMMENTS=${PHP_OPCACHE_SAVE_COMMENTS:="0"}

smalte build --scope PORT --scope NGINX\.* --scope PHP\.* \
    /etc/nginx/host.conf.tmpl:/etc/nginx/host.conf \
    /etc/nginx/nginx.conf.tmpl:/etc/nginx/nginx.conf \
    /etc/php/${PHP_VERSION}/fpm/pool.d/www.conf.tmpl:/etc/php/${PHP_VERSION}/fpm/pool.d/www.conf \
    /etc/php/${PHP_VERSION}/fpm/php.ini.tmpl:/etc/php/${PHP_VERSION}/fpm/php.ini \
    /etc/php/${PHP_VERSION}/cli/php.ini.tmpl:/etc/php/${PHP_VERSION}/cli/php.ini \
    /etc/php/${PHP_VERSION}/fpm/php-fpm.conf.tmpl:/etc/php/${PHP_VERSION}/fpm/php-fpm.conf

echo "> Configuration loaded."
echo "> Run application."

# Start supervisord and services
/usr/local/bin/supervisord -c /etc/supervisord.conf
